sos/report/plugins/openshift.py

```diff
@@ -7,151 +7,369 @@
+    default namespaces/projects that are created during deployment - i.e. the
+    namespaces of the cluster projects matching openshift.* and kube.*. At the
+    time of this plugin's creation that number of default projects is already
+    north of 50; hence this plugin is expected to take a long time in both the
+    setup() and collect() phases. End-user projects may also be collected from
+    when those projects are included in the `add-namespaces` or
+    `only-namespaces` options.
+
+    It is expected to need to perform an `oc login` command in order for this
+    plugin to be able to correctly capture information, as system root is not
+    considered cluster root on the cluster nodes in order to access the API.
+
+    Users will need to either:
+
+        1) Provide the bearer token via the `-k openshift.token` option
+        2) Provide the bearer token via the `SOSOCPTOKEN` environment variable
+        3) Otherwise ensure that the root user can successfully run `oc` and
+           get proper output prior to running this plugin
+
 
-    short_desc = 'Openshift 2.x node and broker'
+    It is highly suggested that option #2 be used first, as this will prevent
+    the token from being recorded in output saved to the archive. Option #1 may
+    be used if this is considered an acceptable risk. It is not recommended to
+    rely on option #3, though it will provide the functionality needed.
+    """
+
+    short_desc = 'Openshift Container Platform 4.x'
 
     plugin_name = "openshift"
-    profiles = ('virt', 'openshift')
+    plugin_timeout = 600
+    profiles = ('openshift',)
+    packages = ('openshift-hyperkube',)
+
+    option_list = [
+        ('token', 'admin token to allow API queries', 'fast', None),
+        ('host', 'host address to use for oc login, including port', 'fast',
+         'https://localhost:6443'),
+        ('no-oc', 'do not collect `oc` command output', 'fast', False),
+        ('podlogs', 'collect logs from each pod', 'fast', True),
+        ('podlogs-filter', ('limit podlogs collection to pods matching this '
+         'regex'), 'fast', ''),
+        ('only-namespaces', 'colon-delimited list of namespaces to collect',
+         'fast', ''),
+        ('add-namespaces', ('colon-delimited list of namespaces to add to the '
+         'default collections'), 'fast', '')
+    ]
+
+    def _check_oc_function(self):
+        """Check to see if we can run `oc` commands"""
+        return self.exec_cmd('oc whoami')['status'] == 0
+
+    def _check_oc_logged_in(self):
+        """See if we're logged in to the API service, and if not attempt to do
+        so using provided plugin options
+        """
+        if self._check_oc_function():
+            return True
+
+        # Not logged in currently, attempt to do so
+        if self.get_option('token') or os.getenv('SOSOCPTOKEN', None):
+            token = self.get_option('token') or os.getenv('SOSOCPTOKEN', None)
```

Comment:
Isn't bit faster to re-use the statement as:

    token = self.get_option('token') or os.getenv('SOSOCPTOKEN', None)
    if token:
        oc_res = ..
        ..

Reply:
Ack.